// @parity/hardhat-polkadot-node/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_TIMEOUT_MILISECONDS = exports.ETH_RPC_TO_SUBSTRATE_RPC = exports.NETWORK_ETH = exports.NETWORK_GAS_PRICE = exports.NETWORK_GAS = exports.POLKADOT_NETWORK_ACCOUNTS = exports.BASE_URL = exports.POLKADOT_TEST_NODE_NETWORK_NAME = exports.NODE_RPC_URL_BASE_URL = exports.RPC_ENDPOINT_PATH = exports.PORT_CHECK_DELAY = exports.MAX_PORT_ATTEMPTS = exports.ETH_RPC_ADAPTER_START_PORT = exports.NODE_START_PORT = exports.PROCESS_TERMINATION_SIGNALS = exports.TASK_RUN_POLKADOT_NODE_IN_SEPARATE_PROCESS = exports.TASK_NODE_POLKADOT_CREATE_ETH_ADAPTER = exports.TASK_NODE_POLKADOT_CREATE_SERVER = exports.TASK_NODE_POLKADOT = exports.PLUGIN_NAME = void 0;
exports.PLUGIN_NAME = "hardhat-polkadot-node";
exports.TASK_NODE_POLKADOT = "node-polkadot";
exports.TASK_NODE_POLKADOT_CREATE_SERVER = "node-polkadot:create-server";
exports.TASK_NODE_POLKADOT_CREATE_ETH_ADAPTER = "node-polkadot:create-eth-adapter";
exports.TASK_RUN_POLKADOT_NODE_IN_SEPARATE_PROCESS = "node-polkadot:run-in-separate-process";
exports.PROCESS_TERMINATION_SIGNALS = ["SIGINT", "SIGTERM", "SIGKILL"];
exports.NODE_START_PORT = 8000;
exports.ETH_RPC_ADAPTER_START_PORT = 8545;
exports.MAX_PORT_ATTEMPTS = 10;
exports.PORT_CHECK_DELAY = 500;
exports.RPC_ENDPOINT_PATH = "eth_chainId";
exports.NODE_RPC_URL_BASE_URL = process.env.CI ? "0.0.0.0" : "host.docker.internal";
exports.POLKADOT_TEST_NODE_NETWORK_NAME = "anvil";
exports.BASE_URL = "http://127.0.0.1";
exports.POLKADOT_NETWORK_ACCOUNTS = [
    "0x5fb92d6e98884f76de468fa3f6278f8807c48bebc13595d45af5bdc4da702133",
    "0x5fb92d6e98884f76de468fa3f6278f8807c48bebc13595d45af5b0000a702133",
    "0x5fb92c48bebcd6e98884f76de468fa3f6278f880713595d45af5b0000a702133",
    "0x5fb92d6de468fa3f6278f8807c48bebc13595d45af5b000e98884f760a702133",
    "0x5fb92d6e98884f76de468fa3f6278f8807c48bebc130a702133595d45af5b000",
    "0x5fb92d6e98884f76de468fa3f6275af5b0000a7021bc338f8807c48be13595d4",
    "0x521338f8807c48befb92d6e98884f76de468fa3f6275af5b0000a70bc13595d4",
    "0x5fb92d6e9884f768de468fa3f6275af5b0000a7021338f8807c48bebc13595d4",
    "0x5fb92d6de468fa3f6275af5b0000a7021338f8807c48bebc13595d46e98884f7",
    "0x5fb92d6e98884f76de468fa3f6275af5b0000a7021338f8807c48be3595d4bc1",
];
exports.NETWORK_GAS = {
    AUTO: "auto",
};
exports.NETWORK_GAS_PRICE = {
    AUTO: "auto",
};
exports.NETWORK_ETH = {
    LOCALHOST: "localhost",
};
exports.ETH_RPC_TO_SUBSTRATE_RPC = {
    "testnet-passet-hub-eth-rpc.polkadot.io": "wss://testnet-passet-hub.polkadot.io",
    "kusama-asset-hub-eth-rpc.polkadot.io": "wss://asset-hub-kusama-rpc.dwellir.com",
    "westend-asset-hub-eth-rpc.polkadot.io": "wss://asset-hub-westend-rpc.dwellir.com",
};
exports.DEFAULT_TIMEOUT_MILISECONDS = 30000;
