// @parity/hardhat-polkadot-node/dist/services/chopsticks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChopsticksService = void 0;
const child_process_1 = require("child_process");
const chalk_1 = __importDefault(require("chalk"));
const constants_1 = require("../constants");
const index_1 = require("./index");
class ChopsticksService extends index_1.Service {
    port;
    constructor(commandArgs = [], blockProcess = true) {
        super(commandArgs, blockProcess);
        const portArg = commandArgs.find((arg) => arg.startsWith("--port="));
        this.port = portArg ? parseInt(portArg.split("=")[1], 10) : constants_1.NODE_START_PORT;
    }
    async from_binary(pathToBinary) {
        return new Promise((resolve, reject) => {
            if (this.blockProcess) {
                console.info(chalk_1.default.green(`Starting server at 127.0.0.1:${this.port}`));
                console.info(chalk_1.default.green(`Running command: ${pathToBinary} ${this.commandArgs.join(" ")}`));
            }
            let stdioConfig = "inherit";
            if (!this.blockProcess) {
                stdioConfig = ["ignore", "ignore", "ignore"];
            }
            this.process = (0, child_process_1.spawn)(this.commandArgs[0], this.commandArgs.slice(1), {
                stdio: stdioConfig,
            });
            this.process.on("error", this._handleOnError("server", reject));
            this.process.on("exit", this._handleOnExit("server"));
            if (!this.blockProcess) {
                resolve();
            }
        });
    }
    async from_docker(_) {
        // not used
    }
    async stop() {
        if (this.process && !this.process.killed) {
            this.process.kill();
        }
    }
}
exports.ChopsticksService = ChopsticksService;
