// @parity/hardhat-polkadot-node/dist/services/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubstrateNodeService = exports.EthRpcService = exports.ChopsticksService = exports.Service = void 0;
const chalk_1 = __importDefault(require("chalk"));
class Service {
    commandArgs;
    blockProcess;
    process = null;
    container = null;
    constructor(commandArgs, blockProcess) {
        this.commandArgs = commandArgs;
        this.blockProcess = blockProcess;
    }
    _handleOnExit(name) {
        return (code, signal) => {
            if (signal) {
                console.info(chalk_1.default.yellow(`Received ${signal} signal. The ${name} process has exited.`));
            }
            else if (code !== 0) {
                console.info(chalk_1.default.red(`The ${name} process exited with code: ${code}`));
            }
        };
    }
    _handleOnError(name, rejectFn) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return (error) => {
            console.info(chalk_1.default.red(`Error running the ${name}:`, error));
            rejectFn(new Error(`Error running the ${name}: ${error.message}`));
        };
    }
}
exports.Service = Service;
var chopsticks_1 = require("./chopsticks");
Object.defineProperty(exports, "ChopsticksService", { enumerable: true, get: function () { return chopsticks_1.ChopsticksService; } });
var eth_rpc_1 = require("./eth-rpc");
Object.defineProperty(exports, "EthRpcService", { enumerable: true, get: function () { return eth_rpc_1.EthRpcService; } });
var substrate_node_1 = require("./substrate-node");
Object.defineProperty(exports, "SubstrateNodeService", { enumerable: true, get: function () { return substrate_node_1.SubstrateNodeService; } });
